### PR TITLE
feat: add support password rotation report

### LINKED ISSUE #105: Add support for the generation of a password rotation report
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

In v1.4.0, we introduced two new cmdlets:

- `Request-PasswordRotationPolicy`
- `Publish-PasswordRotationPolicy`

This enhancement request will track an additional approvement for adding cmdlets to publish a report - `-html` and `-json`  options - for **Password Rotation**. This may be a new report or an addition to the current report (to be decided.)

### Use Case(s)

As a platform engineer, I need to be able to programmatically get the status of the password rotations settings for a credential managed by SDDC Manager.

### Potential Configuration

To be determined.

### References

_No response_

### LINKED ISSUE #109: Update VMware Cloud Foundation version checking
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

In the module, there are 21 instances of the following code:

```powershell
if (((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
    $version = $Matches[0]
}
```

GH-108 introduces the use of the `PowerVCF` v2.4.0 (pre-release). Included in this release is an update to `Get-VCFManager` to return the version. There is an opportunity to use this and reduce the code.

```powershell
PS F:\PowerVCF> Request-VcfToken -fqdn sfo-vcf01.sfo.rainpole.io -username admin@local -password VMw@re1!VMw@re1!
Successfully Requested New API Token From SDDC Manager: sfo-vcf01.sfo.rainpole.io

PS F:\PowerVCF> Get-VCFManager -version
4.5.1.0
```

### Use Case(s)

Technical Debt: There is an opportunity to use this and reduce the code.

### Potential Configuration

```powershell
$version = Get-VCFManager -version
```

### References

_No response_

### PR DESCRIPTION
**Pull request by sowjuec**
### Summary

Added `Invoke-PasswordRotationManager`, `Save-ClarityReportNavigationForRotation` and `Set-CreateReportDirectoryRotation` cmdlets and updated `Publish-PasswordRotationPolicy` cmdlet to generate report for password rotation settings for accounts managed by SDDC Manager in HTML or JSON format. Results can be filtered by workload domain.

Updated `Invoke-PasswordPolicyManager` ,  `Get-PasswordPolicyDefault`, `Get-PasswordPolicyConfig`, `Request-SddcManagerPasswordComplexity`, `Request-SddcManagerAccountLockout`, `Request-SsoAccountLockout`, `Request-VcenterPasswordExpiration`, `Request-VcenterPasswordComplexity`, `Request-VcenterAccountLockout`, `Request-VcenterRootPasswordExpiration`, `Request-NsxtManagerPasswordExpiration`, `Request-NsxtManagerPasswordComplexity`, `Request-NsxtManagerAccountLockout`, `Request-NsxtEdgePasswordExpiration`, `Request-NsxtEdgePasswordComplexity`, `Request-NsxtEdgeAccountLockout`, `Request-EsxiPasswordExpiration`, `Request-EsxiPasswordComplexity`, `Request-EsxiAccountLockout` and `Request-LocalUserPasswordExpiration` cmdlets for simplifying logic for reading version information with PowerVCF v2.4.0 (pending release).

### Type

- [ ] Bugfix
- [X] Enhancement or Feature
- [ ] Code Style or Formatting
- [ ] Documentation
- [ ] Refactoring
- [ ] Chore
- [ ] Other
        Please describe:

### Breaking Changes?

- [ ] Yes, there are breaking changes.
- [X] No, there are no breaking changes.

### Test and Documentation

- [X] Tests have been completed.
- [ ] Documentation has been added or updated.

### Issue References

 Closes #105 
 Closes #109

### Additional Information

Though reporting structure has a mention of 'Aria' suite components. It it yet to be qualified with VCF 5.1.

### CODE REVIEW

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 1]
```diff
@@ -1212,6 +1485,24 @@ Function Set-CreateReportDirectory {
...
 }
 
+Function Set-CreateReportDirectoryRotation {
+    Param (
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$path,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$sddcManagerFqdn
+    )
+
+    $filetimeStamp = Get-Date -Format "MM-dd-yyyy_hh_mm_ss"
+    $Global:reportFolder = $path + '\PasswordRotationManager\'
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:1495]: Could we use `Join-Path` here?

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Noticed that an update to the `mkdocs.yml` is needed to include the cmdlet documentation.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,14 @@
 # Release History
 
+## UnReleased
+
+Enhancement:
+
+- Added `Invoke-PasswordRotationManager`, `Save-ClarityReportNavigationForRotation` and `Set-CreateReportDirectoryRotation` cmdlets and updated `Publish-PasswordRotationPolicy` cmdlet to generate report for password rotation settings for accounts managed by SDDC Manager in HTML or JSON format. Results can be filtered by workload domain. [GH-115](https://github.com/vmware/powershell-module-for-vmware-cloud-foundation-password-management/pull/115) 
+
```
Comment (tenthirtyam) [inline CHANGELOG.md:8]:
```suggestion
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,14 @@
 # Release History
 
+## UnReleased
```
Comment (tenthirtyam) [inline CHANGELOG.md:3]:
```suggestion
## v1.5.0 (Unreleased)
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,14 @@
...
+
+- Updated `Invoke-PasswordPolicyManager`,  `Get-PasswordPolicyDefault`, `Get-PasswordPolicyConfig`, `Request-SddcManagerPasswordComplexity`, `Request-SddcManagerAccountLockout`, `Request-SsoAccountLockout`, `Request-VcenterPasswordExpiration`, `Request-VcenterPasswordComplexity`, `Request-VcenterAccountLockout`, `Request-VcenterRootPasswordExpiration`, `Request-NsxtManagerPasswordExpiration`, `Request-NsxtManagerPasswordComplexity`, `Request-NsxtManagerAccountLockout`, `Request-NsxtEdgePasswordExpiration`, `Request-NsxtEdgePasswordComplexity`, `Request-NsxtEdgeAccountLockout`, `Request-EsxiPasswordExpiration`, `Request-EsxiPasswordComplexity`, `Request-EsxiAccountLockout` and `Request-LocalUserPasswordExpiration` cmdlet simpify logic for reading version information. [GH-115](https://github.com/vmware/powershell-module-for-vmware-cloud-foundation-password-management/pull/115)
```
Comment (tenthirtyam) [inline CHANGELOG.md:9]:
```suggestion
- Updated `Invoke-PasswordPolicyManager`, `Get-PasswordPolicyDefault`, `Get-PasswordPolicyConfig`, `Request-SddcManagerPasswordComplexity`, `Request-SddcManagerAccountLockout`, `Request-SsoAccountLockout`, `Request-VcenterPasswordExpiration`, `Request-VcenterPasswordComplexity`, `Request-VcenterAccountLockout`, `Request-VcenterRootPasswordExpiration`, `Request-NsxtManagerPasswordExpiration`, `Request-NsxtManagerPasswordComplexity`, `Request-NsxtManagerAccountLockout`, `Request-NsxtEdgePasswordExpiration`, `Request-NsxtEdgePasswordComplexity`, `Request-NsxtEdgeAccountLockout`, `Request-EsxiPasswordExpiration`, `Request-EsxiPasswordComplexity`, `Request-EsxiAccountLockout` and `Request-LocalUserPasswordExpiration` cmdlet to simplify the logic for reading version information. [GH-115](https://github.com/vmware/powershell-module-for-vmware-cloud-foundation-password-management/pull/115)
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 5]
```diff
@@ -40,6 +40,281 @@ if ($PSEdition -eq 'Desktop') {
...
+                }
+
+                
+                if ($PsBoundParameters.ContainsKey("json")) {
+
+                    # Collect Password Rotation Setting Data
+                    Write-LogMessage -Type INFO -Message "Collecting SDDC Manager Password Rotation for $workflowMessage."
+                    $sddcManagerPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'backup' $($commandSwitch)" | convertfrom-json
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting Virtual Center Server Single Sign On Rotation for $workflowMessage."
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:140]:
```suggestion
                    Write-LogMessage -Type INFO -Message "Collecting vCenter Server Single Sign-On Rotation for $workflowMessage."
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 6]
```diff
@@ -40,6 +40,281 @@ if ($PSEdition -eq 'Desktop') {
...
+                    $vcenterServerPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'vcenterServer' $($commandSwitch)" | convertfrom-json
+
+                    Write-LogMessage -Type INFO -Message "Collecting NSX Manager Password Rotation for $workflowMessage."
+                    $nsxManagerPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'nsxManager' $($commandSwitch)" | convertfrom-json
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting NSX Edge Password Rotation for $workflowMessage."
+                    $nsxEdgePasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'nsxEdge' $($commandSwitch)" | convertfrom-json
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting WorkSpace One Accesss Password Rotation for $workflowMessage."
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:152]:
```suggestion
                    Write-LogMessage -Type INFO -Message "Collecting WorkSpace ONE Access Password Rotation for $workflowMessage."
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 7]
```diff
@@ -40,6 +40,281 @@ if ($PSEdition -eq 'Desktop') {
...
+                    $jsonFile = ($reportFolder + "passwordRotationManager" + ".json")
+                    Write-LogMessage -Type INFO -Message "Generating the Final JSON and Saving to ($jsonFile)."
+                    $outputJsonObject | ConvertTo-Json -Depth 25| Out-File -FilePath $jsonFile
+                } else {
+
+                    # Collect Password Rotation Setting Data
+                    Write-LogMessage -Type INFO -Message "Collecting SDDC Manager Password Rotation for $workflowMessage."
+                    $sddcManagerPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'backup' $($commandSwitch)" 
+
+                    Write-LogMessage -Type INFO -Message "Collecting Virtual Center Server Single Sign On Rotation for $workflowMessage."
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:223]:
```suggestion
                    Write-LogMessage -Type INFO -Message "Collecting vCenter Server Single Sign-On Rotation for $workflowMessage."
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 8]
```diff
@@ -40,6 +40,281 @@ if ($PSEdition -eq 'Desktop') {
...
+                    $ssoPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'sso' $($commandSwitch)" 
+
+                    Write-LogMessage -Type INFO -Message "Collecting Virtual Center Server Password Rotation for $workflowMessage."
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:226]:
```suggestion
                    Write-LogMessage -Type INFO -Message "Collecting vCenter Server Password Rotation for $workflowMessage."
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 9]
```diff
@@ -40,6 +40,281 @@ if ($PSEdition -eq 'Desktop') {
...
+                    $vcenterServerPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'vcenterServer' $($commandSwitch)" 
+
+                    Write-LogMessage -Type INFO -Message "Collecting NSX Manager Password Rotation for $workflowMessage."
+                    $nsxManagerPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'nsxManager' $($commandSwitch)" 
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting NSX Edge Password Rotation for $workflowMessage."
+                    $nsxEdgePasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'nsxEdge' $($commandSwitch)" 
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting WorkSpace One Accesss Password Rotation for $workflowMessage."
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:235]:
```suggestion
                    Write-LogMessage -Type INFO -Message "Collecting WorkSpace ONE Access Password Rotation for $workflowMessage."
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 10]
```diff
@@ -40,6 +40,281 @@ if ($PSEdition -eq 'Desktop') {
...
+                    $ariaAutomationPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'ariaAutomation' $($commandSwitch)" 
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting Aria Operations Password Rotation for $workflowMessage."
+                    $ariaOperationsPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'ariaOperations' $($commandSwitch)" 
+
+                    Write-LogMessage -Type INFO -Message "Collecting Aria Operations Logs Password Rotation for $workflowMessage."
+                    $ariaOperationsLogsPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'ariaOperationsLogs' $($commandSwitch)" 
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting Aria Lifecycle Password Rotation for $workflowMessage."
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:247]:
```suggestion
                    Write-LogMessage -Type INFO -Message "Collecting Aria Suite Lifecycle Password Rotation for $workflowMessage."
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 11]
```diff
@@ -40,6 +40,281 @@ if ($PSEdition -eq 'Desktop') {
...
+                    $ariaAutomationPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'ariaAutomation' $($commandSwitch)" | convertfrom-json
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting Aria Operations Password Rotation for $workflowMessage."
+                    $ariaOperationsPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'ariaOperations' $($commandSwitch)" | convertfrom-json
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting Aria Operations Logs Password Rotation for $workflowMessage."
+                    $ariaOperationsLogsPasswordRotation = Invoke-Expression "Publish-PasswordRotationPolicy -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -resource 'ariaOperationsLogs' $($commandSwitch)" | convertfrom-json
+                    
+                    Write-LogMessage -Type INFO -Message "Collecting Aria Lifecycle Password Rotation for $workflowMessage."
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:164]:
```suggestion
                    Write-LogMessage -Type INFO -Message "Collecting Aria Suite Lifecycle Password Rotation for $workflowMessage."
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Added some minor comments for changes.

**Comment by tenthirtyam** [comment]
Thanks! I will review very soon.

**Comment by tenthirtyam** [comment]
@sowjuec I don’t see a new or amended commit with the requested changes. Did you push?

**Comment by sowjuec** [comment]
> @sowjuec I don’t see a new or amended commit with the requested changes. Did you push?
@tenthirtyam -- could you please check now?

**Review by tenthirtyam — APPROVED** [review]
LGTM! 🚀

**Pull request merged**